### DISCUSSION TITLE
CacheVertex.refresh() might not refresh cached edges on another server

### DISCUSSION THREAD

**Discussion by porunov** [JanusGraph/janusgraph#3174] [Q&A]
I'm not 100% sure but it looks like `CacheVertex.refresh()` doesn't refresh cached edges related to this vertex for some reason.
It looks like this is reproducible only on separate JanusGraph servers. I started 2 JanusGraph servers which don't share the same JVM and after I added an edge from one vertex to another I can't retrieve this edge on another server even after refresh. 
It could be that I'm doing something wrong. 
@li-boxuan as you were adding this feature, would you be able to confirm that `refresh` method refreshes cache only in a single JanusGraph instance but not on different JanusGraph servers? If that's the case, that's fine and I should probably find out how to invalidate cache properly. Otherwise it could probably be a bug. Invalidating cache in a single JanusGraph instance works as expected. For example, the below test passes because 2 transactions are related to the same JanusGraph instance. 
```
@Test
public void testRelationsCacheForceRefresh() {
    if (features.hasLocking()) return;

    JanusGraphVertex v1 = graph.addVertex();
    JanusGraphVertex v2 = graph.addVertex();

    graph.tx().commit();

    JanusGraphTransaction tx1 = graph.newTransaction();
    assertFalse(tx1.traversal().V(v1.id()).bothE().hasNext());

    JanusGraphTransaction tx2 = graph.newTransaction();
    tx2.traversal().V(v1.id()).addE("relation").to(tx2.traversal().V(v2.id()).next()).iterate();
    tx2.commit();

    // tx1 sees old (cached) edges (i.e. no edges in this case)
    assertFalse(tx1.traversal().V(v1.id()).bothE().hasNext());
    // force refreshing v1 and v2 in tx1, now it should see the new added edge
    ((CacheVertex) tx1.traversal().V(v1.id()).next()).refresh();
    ((CacheVertex) tx1.traversal().V(v2.id()).next()).refresh();

    assertTrue(tx1.traversal().V(v1.id()).bothE().hasNext());

    tx1.commit();
}
```

**Comment by li-boxuan** [comment]
This is weird, could it be because you have database level cache enabled?

**Reply by porunov** [reply]
Ah, sorry, I thought that refresh allows to invalidate vertex with database cache enabled but I see it doesn't. Do you know by any chance if we can invalidate it in the database cache? 
I guess if I find a good way to invalidate cache for anything related to a specific vertex / edge then it would be possible to add additional cache invalidation strategy described in #3155

**Reply by li-boxuan** [reply]
> Do you know by any chance if we can invalidate it in the database cache

Not that I am aware of. The key in the DB cache is a KeySliceQuery. To invalidate a vertex, we need to evict all KeySliceQuery whose `getKey()` match the given vertex.
